src/util/taproot.rs

```diff
@@ -82,12 +93,804 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+
+// type alias for versioned tap script corresponding merkle proof
+type ScriptMerkleProofMap = BTreeMap<(Script, LeafVersion), BTreeSet<TaprootMerkleBranch>>;
+/// Data structure for representing Taproot spending information.
+/// Taproot output corresponds to a combination of a
+/// single public key condition (known the internal key), and zero or more
+/// general conditions encoded in scripts organized in the form of a binary tree.
+///
+/// Taproot can be spent be either:
+/// - Spending using the key path i.e., with secret key corresponding to the output_key
+/// - By satisfying any of the scripts in the script spent path. Each script can be satisfied by providing
+///   a witness stack consisting of the script's inputs, plus the script itself and the control block.
+///
+/// If one or more of the spending conditions consist of just a single key (after aggregation),
+/// the most likely one should be made the internal key.
+/// See [BIP341](https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki) for more details
+/// on choosing internal keys for a taproot application
+///
+/// Note: This library currently does not support [annex](https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki#cite_note-5)
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct TaprootSpendInfo {
+    /// The BIP341 internal key.
+    internal_key: schnorr::PublicKey,
+    /// The Merkle root of the script tree (None if there are no scripts)
+    merkle_root: Option<TapBranchHash>,
+    /// The sign final output pubkey as per BIP 341
+    output_key_parity: bool,
+    /// The tweaked output key
+    output_key: schnorr::PublicKey,
+    /// Map from (script, leaf_version) to (sets of) [`TaprootMerkleBranch`].
+    /// More than one control block for a given script is only possible if it
+    /// appears in multiple branches of the tree. In all cases, keeping one should
+    /// be enough for spending funds, but we keep all of the paths so that
+    /// a full tree can be constructed again from spending data if required.
+    script_map: ScriptMerkleProofMap,
+}
+
+impl TaprootSpendInfo {
+    /// Create a new [`TaprootSpendInfo`] from a list of script(with default script version) and
+    /// weights of satisfaction for that script. The weights represent the probability of
+    /// each branch being taken. If probabilities/weights for each condition are known,
+    /// constructing the tree as a Huffman tree is the optimal way to minimize average
+    /// case satisfaction cost. This function takes input a list of tuple(usize, &Script)
+    /// where usize represents the satisfaction weights of the branch.
+    /// For example, [(3, S1), (2, S2), (5, S3)] would construct a TapTree that has optimal
+    /// satisfaction weight when probability for S1 is 30%, S2 is 20% and S3 is 50%.
+    ///
+    /// # Errors:
+    ///
+    /// - When the optimal huffman tree has a depth more than 128
```

Comment:
This is an improvement. I can create a good first issue for this. If someone really wants this, they are welcome to implement this and I am happy to review it.

Reply:
If the optimal huffman tree has depth exceeding 128 it means the user has set insane weights. We should error rather than silently doing the wrong thing.